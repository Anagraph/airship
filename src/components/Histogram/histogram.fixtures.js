export default [
  {
    bin: 0,
    start: 9,
    end: 32.3333333333333,
    freq: 1667,
    min: 9,
    max: 32,
    avg: 24.083983203359328,
    normalized: 0.7874350495984884,
  },
  {
    bin: 1,
    start: 32.3333333333333,
    end: 55.6666666666666,
    freq: 2117,
    min: 33,
    max: 55,
    avg: 44.31979215871516,
    normalized: 1,
  },
  {
    bin: 2,
    start: 55.6666666666666,
    end: 78.9999999999999,
    freq: 1638,
    min: 56,
    max: 78,
    avg: 66.23076923076923,
    normalized: 0.7737364194615022,
  },
  {
    bin: 3,
    start: 78.9999999999999,
    end: 102.3333333333332,
    freq: 1071,
    min: 79,
    max: 102,
    avg: 89.14472455648927,
    normalized: 0.5059045819555975,
  },
  {
    bin: 4,
    start: 102.3333333333332,
    end: 125.6666666666665,
    freq: 337,
    min: 103,
    max: 125,
    avg: 116.11572700296736,
    normalized: 0.15918752952290977,
  },
  {
    bin: 5,
    start: 125.6666666666665,
    end: 149,
    freq: 180,
    min: 127,
    max: 149,
    avg: 136.61666666666667,
    normalized: 0.08502598016060463,
  },
];
