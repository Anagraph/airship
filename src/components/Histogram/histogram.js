import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { darken } from 'polished';
import { select, selectAll } from 'd3-selection'; // eslint-disable-line
import { scaleBand, scaleLinear } from 'd3-scale';
import { min, max } from 'd3-array'; // eslint-disable-line
import { transition } from 'd3-transition'; // eslint-disable-line
import { axisBottom, axisLeft } from 'd3-axis'; // eslint-disable-line

const Svg = styled.svg`
  max-width: 248px;

  .tick {
    line {
      opacity: 0.1;
    }

    text {
      color: ${props => props.textColor};
    }

    :first-child {
      line {
        opacity: 0.5;
      }
    }
  }

  .bar {
    fill: ${props => props.color};
    cursor: pointer;
    opacity: 0.7;
    transition: all 0.3s linear;

    :hover {
      fill: ${props => darken(0.16, props.color)};
    }
  }
`;

class Histogram extends Component {
  static defaultProps = {
    color: '#3AB5F0',
    data: [],
    textColor: '#747474',
    width: 208,
    height: 140,
    margin: {
      top: 15,
      right: 0,
      bottom: 5,
      left: 40,
    },
  };

  static propTypes = {
    color: PropTypes.string,
    data: PropTypes.array,
    height: PropTypes.number,
    margin: PropTypes.object,
    textColor: PropTypes.string,
    width: PropTypes.number,
  };

  componentDidMount() {
    const { margin } = this.props;

    this.barsContainer = this.container
      .append('g')
      .attr('transform', `translate(${margin.left}, ${margin.top})`);

    this.renderAxis();
    this.renderBars();
  }

  componentDidUpdate() {
    this.updateAxis();
    this.renderBars();
  }

  updateAxis() {
    const { data } = this.props;

    this.yScale
      .domain([0, max(data, d => d.value)]);

    this.xScale
      .domain(data.map(d => d.name));

    this.yAxisSelection
      .call(this.yAxis);

    select('.domain').remove(); // Remove axis border
  }

  renderAxis() {
    const { data, height, margin, width } = this.props;

    // -- X Axis
    this.xScale = scaleBand()
      .paddingInner(0.05)
      .paddingOuter(0.1)
      .domain(data.map(d => d.name))
      .range([0, width]);

    // -- Y Axis
    this.yScale = scaleLinear()
      .domain([0, max(data, d => d.value)])
      .range([height, 0]);

    this.yAxis = axisLeft(this.yScale)
      .tickSize(-width, 0, 0)
      .ticks(5);

    this.yAxisSelection = this.container
      .append('g')
      .attr('transform', `translate(${margin.left}, ${margin.top})`)
      .call(this.yAxis);

    select('.domain').remove(); // Remove axis border
  }

  renderBars() {
    const { data, height } = this.props;

    // Draw bars
    this.bars = this.barsContainer
      .selectAll('rect')
      .data(data);

    // Exit
    this.bars.exit().remove();

    // Enter
    this.bars
      .enter()
      .append('rect')
      .merge(this.bars)
      .attr('class', 'bar')
      .attr('y', height)
      .attr('x', d => this.xScale(d.name))
      .attr('width', () => this.xScale.bandwidth())
      .transition()
      .delay(200)
      .attr('y', d => this.yScale(d.value))
      .attr('height', d => height - this.yScale(d.value));

    // Update
    this.bars
      .attr('y', d => this.yScale(d.value))
      .attr('height', d => height - this.yScale(d.value));
  }

  render() {
    const { width, height, margin, ...others } = this.props;
    const fullHeight = height + margin.top + margin.bottom;
    const fullWidth = width + margin.left + margin.right;

    return (
      <Svg
        width={fullWidth}
        height={fullHeight}
        innerRef={node => { this.container = select(node); }}
        {...others}
      />
    );
  }
}

export default Histogram;
