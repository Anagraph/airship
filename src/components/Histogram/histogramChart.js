import { select, selectAll } from 'd3-selection';
import { scaleBand, scaleLinear } from 'd3-scale';
import { rgb } from 'd3-color';
import { min, max } from 'd3-array';
import { axisBottom, axisLeft } from 'd3-axis';
import { readableNumber, truncate } from '../../utils';

const DEFAULT_OPTIONS = {
  width: 208,
  height: 140,
  margin: {
    top: 15,
    right: 0,
    bottom: 5,
    left: 40,
  },
};

export default class Gaugechart {
  constructor(element, options) {
    if (!element) throw new Error('A root container is required');

    this.element = element;
    this.options = Object.assign({}, DEFAULT_OPTIONS, options);

    this.createChart();
  }

  createChart() {
    const { data: { bins }, height, margin, width } = this.options;
    const fullHeight = height + margin.top + margin.bottom;
    const fullWidth = width + margin.left + margin.right;

    // -- Base SVG
    const svg = select(this.element)
      .append('svg')
        .attr('width', fullWidth)
        .attr('height', fullHeight)
      .append('g')
        .attr('transform', `translate(${margin.left}, ${margin.top})`);

    // -- X Axis
    const xScale = scaleBand()
      .paddingInner(0.05)
      .paddingOuter(0.1)
      .domain(bins.map(d => d.max))
      .range([0, width]);

    // -- Y Axis
    const yScale = scaleLinear()
      .domain([0, max(bins, d => d.freq)])
      .range([height, 0]);

    const yAxis = axisLeft(yScale)
      .tickSize(-width, 0, 0)
      .ticks(5);

    svg.call(yAxis);

    select('.domain').remove();

    // -- Create bars
    svg.selectAll('rect')
      .data(bins)
      .enter()
      .append('rect')
        .attr('class', 'bar')
        .attr('x', d => xScale(d.max))
        .attr('y', d => yScale(d.freq))
        .attr('width', d => xScale.bandwidth())
        .attr('height', d => yScale(0) - yScale(d.freq))


    // this._refresh();
  }

  update(options) {
    this.options = { ...this.options, ...options };

    this.createChart(); // TODO: Replace with refresh and animations
  }

  _refresh() {
    const { data, colors, showLegend } = this.options;
  }
}
