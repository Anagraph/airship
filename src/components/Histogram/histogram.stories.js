import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import Histogram from './histogram';
import Widget from '../Widget/widget';
import mockData from './histogram.fixtures';

class HistogramUpdated extends React.Component {
  state = {
    data: mockData
  }

  changeState = () => {
    this.setState(prevState => ({
      data: prevState.data.length !== 4 ? mockData.slice(0, 4) : mockData
    }))
  }

  render() {
    return (
      <div>
        <Histogram data={this.state.data} />
        <button onClick={this.changeState}>Click me</button>
      </div>
    )
  }
}

storiesOf('Histogram', module)
  .add('Default', () => (
    <Histogram data={mockData} />
  ))
  .add('Inside a widget', () => (
    <Widget>
      <Widget.Title>Suffer score</Widget.Title>
      <Widget.Description>Just a widget</Widget.Description>

      <Histogram data={mockData} />
    </Widget>
  ))
  .add('Updating data', () => (
    <Widget>
      <Widget.Title>Suffer score</Widget.Title>
      <Widget.Description>Just a widget</Widget.Description>

      <HistogramUpdated />
    </Widget>
  ));
