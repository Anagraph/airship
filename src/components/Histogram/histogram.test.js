import React from 'react';
import renderer from 'react-test-renderer';
import Histogram from './histogram';
import mockData from './histogram.fixtures'


describe('<Histogram />', () => {
  describe('render', () => {
    describe('when data passed in', () => {
      it('renders the correct amount of bars', () => {
        const component = renderer.create(<Histogram data={mockData} />);
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe('update', () => {
    describe('when data changes', () => {
      it('should call the update method or the chart', () => {
        const component = renderer.create(<Histogram data={mockData} />);
        const tree = component.toJSON();

        expect(component.toJSON()).toMatchSnapshot();

        // Changing properties should trigger a componentDidUpdate
        component.update(<Histogram data={[]} />);

        expect(component.toJSON()).toMatchSnapshot();
      });
    });
  });
});
