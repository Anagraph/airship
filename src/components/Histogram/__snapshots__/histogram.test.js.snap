// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Histogram /> render when data passed in renders the correct amount of bars 1`] = `
.c0 {
  max-width: 248px;
}

.c0 .tick line {
  opacity: 0.3;
}

.c0 .tick text {
  color: #747474;
}

.c0 .tick:first-child line {
  opacity: 0.5;
}

.c0 .bar {
  fill: #3AB5F0;
  cursor: pointer;
  opacity: 0.7;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
}

.c0 .bar:hover {
  fill: #0f8dc9;
}

<svg
  className="c0"
  data={
    Array [
      Object {
        "avg": 24.083983203359328,
        "bin": 0,
        "end": 32.3333333333333,
        "freq": 1667,
        "max": 32,
        "min": 9,
        "normalized": 0.7874350495984884,
        "start": 9,
      },
      Object {
        "avg": 44.31979215871516,
        "bin": 1,
        "end": 55.6666666666666,
        "freq": 2117,
        "max": 55,
        "min": 33,
        "normalized": 1,
        "start": 32.3333333333333,
      },
      Object {
        "avg": 66.23076923076923,
        "bin": 2,
        "end": 78.9999999999999,
        "freq": 1638,
        "max": 78,
        "min": 56,
        "normalized": 0.7737364194615022,
        "start": 55.6666666666666,
      },
      Object {
        "avg": 89.14472455648927,
        "bin": 3,
        "end": 102.3333333333332,
        "freq": 1071,
        "max": 102,
        "min": 79,
        "normalized": 0.5059045819555975,
        "start": 78.9999999999999,
      },
      Object {
        "avg": 116.11572700296736,
        "bin": 4,
        "end": 125.6666666666665,
        "freq": 337,
        "max": 125,
        "min": 103,
        "normalized": 0.15918752952290977,
        "start": 102.3333333333332,
      },
      Object {
        "avg": 136.61666666666667,
        "bin": 5,
        "end": 149,
        "freq": 180,
        "max": 149,
        "min": 127,
        "normalized": 0.08502598016060463,
        "start": 125.6666666666665,
      },
    ]
  }
  height={160}
  width={248}
/>
`;

exports[`<Histogram /> update when data changes should call the update method or the chart 1`] = `
.c0 {
  max-width: 248px;
}

.c0 .tick line {
  opacity: 0.3;
}

.c0 .tick text {
  color: #747474;
}

.c0 .tick:first-child line {
  opacity: 0.5;
}

.c0 .bar {
  fill: #3AB5F0;
  cursor: pointer;
  opacity: 0.7;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
}

.c0 .bar:hover {
  fill: #0f8dc9;
}

<svg
  className="c0"
  data={
    Array [
      Object {
        "avg": 24.083983203359328,
        "bin": 0,
        "end": 32.3333333333333,
        "freq": 1667,
        "max": 32,
        "min": 9,
        "normalized": 0.7874350495984884,
        "start": 9,
      },
      Object {
        "avg": 44.31979215871516,
        "bin": 1,
        "end": 55.6666666666666,
        "freq": 2117,
        "max": 55,
        "min": 33,
        "normalized": 1,
        "start": 32.3333333333333,
      },
      Object {
        "avg": 66.23076923076923,
        "bin": 2,
        "end": 78.9999999999999,
        "freq": 1638,
        "max": 78,
        "min": 56,
        "normalized": 0.7737364194615022,
        "start": 55.6666666666666,
      },
      Object {
        "avg": 89.14472455648927,
        "bin": 3,
        "end": 102.3333333333332,
        "freq": 1071,
        "max": 102,
        "min": 79,
        "normalized": 0.5059045819555975,
        "start": 78.9999999999999,
      },
      Object {
        "avg": 116.11572700296736,
        "bin": 4,
        "end": 125.6666666666665,
        "freq": 337,
        "max": 125,
        "min": 103,
        "normalized": 0.15918752952290977,
        "start": 102.3333333333332,
      },
      Object {
        "avg": 136.61666666666667,
        "bin": 5,
        "end": 149,
        "freq": 180,
        "max": 149,
        "min": 127,
        "normalized": 0.08502598016060463,
        "start": 125.6666666666665,
      },
    ]
  }
  height={160}
  width={248}
/>
`;

exports[`<Histogram /> update when data changes should call the update method or the chart 2`] = `
.c0 {
  max-width: 248px;
}

.c0 .tick line {
  opacity: 0.3;
}

.c0 .tick text {
  color: #747474;
}

.c0 .tick:first-child line {
  opacity: 0.5;
}

.c0 .bar {
  fill: #3AB5F0;
  cursor: pointer;
  opacity: 0.7;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
}

.c0 .bar:hover {
  fill: #0f8dc9;
}

<svg
  className="c0"
  data={Array []}
  height={160}
  width={248}
/>
`;
